File: 0155-min-stack/0155-min-stack.py
"""
class MinStack:

    def __init__(self):
        self.min_val = 0
        self.stack = []

    def push(self, val: int) -> None:
        if len(self.stack) == 0 or val < self.min_val:
            self.min_val = val
        self.stack.append(val)    

    def pop(self) -> None:
        val = self.stack.pop()
        if val == self.min_val:
            self._set_next_min()

    def top(self) -> int:
        return self.stack[-1]
        
    def getMin(self) -> int:
        return self.min_val
    
    def _set_next_min(self) -> int:
        if len(self.stack) != 0:
            self.min_val = self.stack[0]
            for n in self.stack:
                self.min_val = min(self.min_val, n)
"""

class MinStack:

    def __init__(self):
        # stack of (min, cur) tuples
        self.stack = []

    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            self.stack.append((val, val))
        else:
            prev_min = self.stack[-1][0]
            new_min = min(prev_min, val)
            self.stack.append((new_min, val))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][1]
        
    def getMin(self) -> int:
        return self.stack[-1][0]

File: 0155-min-stack/0155-min-stack.go
type MinStack struct {
    n *node 
}

type node struct {
    prev *node
    localMin int
    val int
}


func Constructor() MinStack {
    return MinStack{}
}


func (this *MinStack) Push(val int)  {
    if this.n == nil {
        this.n = &node{
            localMin: val,
            val: val,
        }
    } else {
        this.n = &node{
            prev: this.n,
            localMin: min(this.n.localMin, val),
            val: val,
        }
    }
}


func (this *MinStack) Pop() {
    this.n = this.n.prev
}


func (this *MinStack) Top() int {
    return this.n.val
}


func (this *MinStack) GetMin() int {
    return this.n.localMin
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */